### ISSUE TITLE
heatmap slightly misaligned with respect to axis coordinates

### ISSUE THREAD

**Issue by dylanfesta** [MakieOrg/Makie.jl#4053]
I was plotting some analytic curves on top of a heatmap using CairoMackie, and I found out that possibly **all heatmaps are slightly misaligned with respect to the axis coordinates**.  This can be fixed by shifting the heatmap of a factor that I had to find by trial and error (and that  might vary!).  In the example below I generate heatmaps of random numbers, and then draw edges as vertical and horizontal lines. The contours are expected to match.

```julia
using Makie,CairoMakie
using StatsBase: midpoints

"""
  get_edges(v::AbstractVector{R}) where R<:Real

  Given a vector of values corresponding to center coordinates of size N,
  returns a vector of edges N+1.
"""
function get_edges(v::AbstractVector{R}) where R<:Real
  ret = midpoints(v)
  first_edge = v[1] - (ret[1]-v[1])
  last_edge = v[end] + (v[end]-ret[end])
  return vcat(first_edge,ret,last_edge)
end

N=50
fig = Figure(size=(700,300))
xedges = get_edges(1:N) 
yedges = get_edges(1:N)
ax1=Axis(fig[1,1];limits=(extrema(xedges),extrema(yedges)))

heatmap!(ax1,xedges,yedges, randn(N,N))
vlines!(ax1,xedges,linewidth=0.1,color=:white)
hlines!(ax1,yedges,linewidth=0.1,color=:white)
ax2=Axis(fig[1,2]; limits=(extrema(xedges),extrema(yedges)))
xedges_fix = xedges .+ N/500 # this fixes the problem by shifting the heatmap
yedges_fix = yedges .+ N/500 
heatmap!(ax2,xedges_fix,yedges_fix, randn(N,N))
vlines!(ax2,xedges,linewidth=0.1,color=:white)
hlines!(ax2,yedges,linewidth=0.1,color=:white)

colgap!(fig.layout,10)

save("/tmp/tmp.pdf",fig; pt_per_unit=1)

display(fig)
```
This is the plot I see.

![image](https://github.com/user-attachments/assets/7f7a293b-e5b1-4ab1-aab0-75257beddf24)

When I zoom in, the left plot shows a clear misalignment between the heatmap edges and the lines in axis coordinates. In the right plot the misalignment is fixed by adding the `N/500` correction. 

![image](https://github.com/user-attachments/assets/6bea781b-9bf0-4f12-8619-1d60542da48d)

I am using the latest version of Makie and the same problem appears consistently on Ubuntu 20.04 and Fedora 40 machines.

**Comment by jkrumbiegel** [comment]
Hm you're right I also see that in CairoMakie, but not in GLMakie. So that suggests that the tile overlapping that CairoMakie does (to remove antialiasing fringes between heatmap cells) is buggy and we actually see the overlaps, which shouldn't happen. That's why the factor you find is variable, because it's in screen space and not data space.

**Comment by asinghvi17** [comment]
I think we did this deliberately at some stage to prevent aliasong, but maybe the order in which the rectangles are plotted is incorrect, resulting in bad overdraw behavior.

**Comment by dylanfesta** [comment]
Until the bug is fixed, is there a way to know the exact offset value? Can I extract the length of  the axis in pixels? I checked the documentation but I could not find an answer.

**Comment by lanceXwq** [comment]
I encountered a similar misalignment issue when plotting heat maps in an Axis3 using CarioMakie. However, randomly, I found setting alpha to be less than one seemed to fix this issue. I modified the example @dylanfesta provided above such that the second heatmap is now plotted with `alpha=0.999` and the unshifted edges: 
```julia
heatmap!(ax2,xedges,yedges, randn(N,N),alpha=0.999)
vlines!(ax2,xedges,linewidth=0.1,color=:white)
hlines!(ax2,yedges,linewidth=0.1,color=:white)
```
And this is what I get.
![Image](https://github.com/user-attachments/assets/ae91b00d-b9fb-4c54-bbe4-5c5afd81ad1e)

I hope what I found provides a temporary solution and sheds some light on the source of this problem. :)
